### PR TITLE
Open file refactor

### PR DESCRIPTION
**Pull request by RoiEXLab**
Based on #1108
Needs some testing (works on my machine without any problems, but #828 / #862 reminded me of potential problems with this change).
In theory this should work for 99,9% of all java versions

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/1110)

### CODE REVIEW

File: src/games/strategy/net/OpenFileUtility.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      } catch (IOException e) {
+        ClientLogger.logError(e);
+      }
+    } else {
+      logDesktopAPIMessage("URL");
+    }
+  }
+
+  private static void logDesktopAPIMessage(String type) {
+    ClientLogger.logQuietly("The Client doesn't support the Desktop API");
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: If we're going to alert the user on the next line in a message dialog, the print to the error console is redundant.

File: src/games/strategy/net/OpenFileUtility.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    JOptionPane.showMessageDialog(null,
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: Kindly check in SwingComponents if there is a wrapper API for this Swing method (that avoids the useless first `null` arg). If not, please feel free to create one

File: src/games/strategy/net/OpenFileUtility.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+   * Opens the specified web page in the user's default browser
+   *
+   * @param url An URL of a web page (ex: "http://www.google.com/")
+   */
+  public static void openURL(final String url) {
+    if (Desktop.isDesktopSupported()) {
+      try {
+        Desktop.getDesktop().browse(URI.create(url));
+      } catch (IOException e) {
+        ClientLogger.logError(e);
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: Please include the URL in the error message, ie: logError("Failed to open: " + url, e);
Same for above with File method

File: src/games/strategy/net/OpenFileUtility.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+   */
+  public static void openFile(final File file) {
+    if (Desktop.isDesktopSupported()) {
+      try {
+        Desktop.getDesktop().open(file);
+      } catch (IOException e) {
+        ClientLogger.logError(e);
+      }
+    } else {
+      logDesktopAPIMessage("File");
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: Consider doing this logging in the `isDesktopSupported()` method.

File: src/games/strategy/net/OpenFileUtility.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import java.net.URI;
+
+import javax.swing.JOptionPane;
+
+import games.strategy.debug.ClientLogger;
+
+public class OpenFileUtility {
+
+  /**
+   * Opens a specific file on the user's computer, using whatever default program is used to open such files, using the
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: This comment can be tightened up, you lose no information if you delete "using whatever default program is used to open such files, "

IE:  "Opens a specific file on the user's computer using the local computer's file associations."

File: src/games/strategy/net/OpenFileUtility.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import java.awt.Desktop;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+
+import javax.swing.JOptionPane;
+
+import games.strategy.debug.ClientLogger;
+
+public class OpenFileUtility {
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: public classes should have a class comment so the intent of the class does not have to be reverse engineered from the code.  The previous one was overly worded and had unnecessary information, but a few words could be of some help to ensure there is understanding on why we would want to use this.

**Comment by DanVanAtta** [comment]
tested both the open URL and open file parts, seems to work fine with the refactor. Some small cleanup/documentation, mostly reviewed, looking good.

File: src/games/strategy/net/OpenFileUtility.java (hunk 4 above)
Comment (RoiEXLab) [inline src/games/strategy/net/OpenFileUtility.java]: The basic Idea gehind this that if the DesktopAPI is supported, this method should execute without any exceptions - thats why we are logging them - they shouldn't happen.
When java tells us that the DesktopAPI is not supported, then our user has bad luck and has to open the file manually or type in the URL manually - Displaying the Path is a very good suggestion...

File: src/games/strategy/net/OpenFileUtility.java (hunk 2 above)
Comment (RoiEXLab) [inline src/games/strategy/net/OpenFileUtility.java]: Well the first argument is the Parent component - This has to be null unless we want to create a new frame or pass the current one...

**Comment by RoiEXLab** [comment]
@DanVanAtta updated!

File: src/games/strategy/net/OpenFileUtility.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: yes, but we don't have to call this terrible swing API.
Instead the suggestion is to create a static method in SwingComponents if one is not already there, that accepts a paramter and defaults the null for us. In that case the only thing passing null is the SwingComponents implementation.

File: src/games/strategy/net/OpenFileUtility.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: You can still do the logging in 'isDestopSupported()', in that case the clients don't all have to do the logging.

Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: It also let's you get rid of the magic strings "File" and "URL"

File: src/games/strategy/net/OpenFileUtility.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      } catch (IOException e) {
+        ClientLogger.logError("Could not open URL " + url, e);
+      }
+    } else {
+      logDesktopAPIMessage("URL", url);
+    }
+  }
+
+  private static void logDesktopAPIMessage(String type, String path) {
+    ClientLogger.logQuietly("The Client doesn't support the Desktop API");
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: Still have the double error logging. The point of ClientLogger.logQuietly is to suppress notification from the user,  yet the next line does notification to the user. To boot, this message appearing in the error console is likely to yield any additoinal information.

File: src/games/strategy/net/OpenFileUtility.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    JOptionPane.showMessageDialog(null,
+        "We're sorry, but it seems that your installed java version doesn't support the Desktop API required to open the destination "
+            + type + ": " + path);
```
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: The 'type' should be pretty obvious in the error message given the path. The 'type' is not needed.

File: src/games/strategy/net/OpenFileUtility.java (hunk 7 above)
Comment (DanVanAtta) [inline src/games/strategy/net/OpenFileUtility.java]: Explicitly, either drop the ClientLogger.logQuietly, or use ClientLogger.logError and drop the JOptionP:ane.showMessageDialog

File: src/tools/map/making/MapCreator.java [hunk 9]
```diff
@@ -81,6 +81,8 @@ public static void main(final String[] args) {
     creator.setSize(800, 600);
     creator.setLocationRelativeTo(null);
     creator.setVisible(true);
+    OpenFileUtility.openFile(new File("C:\\Users\\Robin\\Desktop\\index.php"));
```
Comment (DanVanAtta) [inline src/tools/map/making/MapCreator.java]: commit leak?

Comment (RoiEXLab) [inline src/tools/map/making/MapCreator.java]: Yes, oooops that was some testing

**Comment by RoiEXLab** [comment]
@DanVanAtta Updated!

> You can still do the logging in 'isDestopSupported()', in that case the clients don't all have to do the logging. 

Not sure how you meant that, but it should be OK now...

**Comment by DanVanAtta** [comment]
thanks for the updates, looks good.

I thought `isDesktopSupported()` was our method, but nope, is a lib.. was mistaken there.

**Pull request merged**